File: qif4j/src/main/java/com/github/squirrelgrip/qif4j/write/QifInvestmentRecord.java
package com.github.squirrelgrip.qif4j.write;

import com.github.squirrelgrip.qif4j.QifInvestment;

public class QifInvestmentRecord extends AbstractQifRecord {
	private QifInvestment txn;

	public QifInvestmentRecord(QifInvestment txn) {
		this.txn = txn;
	}

	@Override
	public QifHeaderEnum getHeader() {
		return QifHeaderEnum.INVESTMENT;
	}

	@Override
	public String formatTransaction() {
		String txnString = null;
		QifInvestmentAction action = QifInvestmentAction.forText(txn
				.getAction());
		switch (action) {
		case DIV:
			txnString = formatDividendTransaction();
			break;
		case TRANSFER_IN:
		case TRANSFER_OUT:
			txnString = formatTransferTransaction();
			break;
		default:
			txnString = formatDefaultTransaction();
			break;
		}
		return txnString;
	}

	private String formatTransferTransaction() {
		StringBuilder sb = new StringBuilder();
		Utils.appendFieldValueToOutput("D", txn.getDate(), sb);
		Utils.appendFieldValueToOutput("N", txn.getAction(), sb);
		Utils.appendFieldValueToOutput("P", "Transfer", sb);
		Utils.appendCurrencyValueToOutput("T", txn.getTotal(), sb);
		Utils.appendFieldValueToOutput("L", txn.getCategory(), sb);
		Utils.appendCurrencyValueToOutput("$", txn.getTotal(), sb);
		return sb.toString();
	}

	private String formatDefaultTransaction() {
		StringBuilder sb = new StringBuilder();
		Utils.appendFieldValueToOutput("D", txn.getDate(), sb);
		Utils.appendFieldValueToOutput("N", txn.getAction(), sb);
		Utils.appendFieldValueToOutput("Y", txn.getSecurity(), sb);
		Utils.appendCurrencyValueToOutput("I", txn.getPrice(), sb);
		Utils.appendFieldValueToOutput("Q", txn.getQuantity(), sb);
		Utils.appendCurrencyValueToOutput("T", txn.getTotal(), sb);
		Utils.appendCurrencyValueToOutput("O", txn.getCommission(), sb);
		Utils.appendFieldValueToOutput("M", txn.getMemo(), sb);
		return sb.toString();
	}

	private String formatDividendTransaction() {
		StringBuilder sb = new StringBuilder();
		Utils.appendFieldValueToOutput("D", txn.getDate(), sb);
		Utils.appendFieldValueToOutput("N", txn.getAction(), sb);
		Utils.appendFieldValueToOutput("Y", txn.getSecurity(), sb);
		Utils.appendCurrencyValueToOutput("I", txn.getPrice(), sb);
		Utils.appendCurrencyValueToOutput("T", txn.getTotal(), sb);
		Utils.appendFieldValueToOutput("M", txn.getMemo(), sb);
		return sb.toString();
	}

}

File: qif4k/src/main/kotlin/com/github/squirrelgrip/qif4k/write/QifInvestmentRecord.kt
package com.github.squirrelgrip.qif4k.write

import com.github.squirrelgrip.qif4k.QifInvestment

class QifInvestmentRecord(private val txn: QifInvestment) : AbstractQifRecord() {
    override val header: QifHeaderEnum
        get() = QifHeaderEnum.INVESTMENT

    override fun formatTransaction(): String? {
        var txnString: String? = null
        val action: QifInvestmentAction = QifInvestmentAction.forText(txn.action)
        txnString = when (action) {
            QifInvestmentAction.DIV -> formatDividendTransaction()
            QifInvestmentAction.TRANSFER_IN, QifInvestmentAction.TRANSFER_OUT -> formatTransferTransaction()
            else -> formatDefaultTransaction()
        }
        return txnString
    }

    private fun formatTransferTransaction(): String {
        val sb = StringBuilder()
        Utils.appendFieldValueToOutput("D", txn.date, sb)
        Utils.appendFieldValueToOutput("N", txn.action, sb)
        Utils.appendFieldValueToOutput("P", "Transfer", sb)
        Utils.appendCurrencyValueToOutput("T", txn.total, sb)
        Utils.appendFieldValueToOutput("L", txn.category, sb)
        Utils.appendCurrencyValueToOutput("$", txn.total, sb)
        return sb.toString()
    }

    private fun formatDefaultTransaction(): String {
        val sb = StringBuilder()
        Utils.appendFieldValueToOutput("D", txn.date, sb)
        Utils.appendFieldValueToOutput("N", txn.action, sb)
        Utils.appendFieldValueToOutput("Y", txn.security, sb)
        Utils.appendCurrencyValueToOutput("I", txn.price, sb)
        Utils.appendFieldValueToOutput("Q", txn.quantity, sb)
        Utils.appendCurrencyValueToOutput("T", txn.total, sb)
        Utils.appendCurrencyValueToOutput("O", txn.commission, sb)
        Utils.appendFieldValueToOutput("M", txn.memo, sb)
        return sb.toString()
    }

    private fun formatDividendTransaction(): String {
        val sb = StringBuilder()
        Utils.appendFieldValueToOutput("D", txn.date, sb)
        Utils.appendFieldValueToOutput("N", txn.action, sb)
        Utils.appendFieldValueToOutput("Y", txn.security, sb)
        Utils.appendCurrencyValueToOutput("I", txn.price, sb)
        Utils.appendCurrencyValueToOutput("T", txn.total, sb)
        Utils.appendFieldValueToOutput("M", txn.memo, sb)
        return sb.toString()
    }

}